### DISCUSSION TITLE
How to get traffic light color information

### DISCUSSION THREAD

**Discussion by ltj3334** [carla-simulator/carla#5508] [Q&A] [answered]
I am studying object detection in CARLA simulation.

My coding skills are not good enough, so I wrote a question.

I had to figure out the color of the traffic light to make the training data, so I succeeded in implementing the Binding Box of the traffic light by following the URL below, but I couldn't figure out the color of the traffic light.

 - URL : https://carla.readthedocs.io/en/latest/tuto_G_bounding_boxes/#bounding-boxes_1

There are two ways I've tried.

1. Obtain information from traffic lights recognized by the Actor Class
   - Result : Failed
   - The traffic light implemented as an Actor object ('traffic.traffic_light') does not have color information.

2. 'is_at_traffic_light(), get_traffic_light()' method
   - Result : Failed
   - As instructed on the homepage, only the red light was returned, so the green light and yellow light could not be identified.
   ```
   if vehicle.is_at_traffic_light(): 
     traffic_light = vehicle.get_traffic_light()
     print(traffic_light.get_state())
   ```

If possible, I would like to receive the traffic light by Actor Class and know the color information

Could you tell me any useful information?

Thank you for reading it.

**Comment by KevinLADLee** [comment, marked as answer]
Hi @ltj3334 , I've done some training data collection job before. You could have a look at this [link](https://github.com/KevinLADLee/carla_dataset_tools/blob/afc43efc9f18472b726261e6a5e9a2aadd3b2c4e/label_tools/yolov5/yolov5_helper.py#L62).

I use a RGB camera and a segmentation camera to extract every light box part in the RGB image, and then determine the color of the light box using some classical method. It is not an elegant way to do this job, but it may helps you quickly collect some training data and quickly to start you further project.

For further developing, you could refer this [link](https://github.com/KevinLADLee/carla_dataset_tools/blob/afc43efc9f18472b726261e6a5e9a2aadd3b2c4e/data_recorder.py#L44-L47) to get all traffic light actors, and use `get_state` to get their state to determine the traffic light color. But I think maybe you want to label the light box, not the traffic light pole. So that you could use the `get_light_boxes` function to get the bbox of light box and project it to the 2D image.

**Reply by ltj3334** [reply]
Thank you for the detailed information.
I think I was trying to solve it too simply.
I read the guide and code you wrote and knew that how to solve it.

Thank you very much for your kind reply! :)

**Comment by SExpert12** [comment]
Hi @ltj3334,
can you please share your code?
I am using TM with traffic light and I am not able to make the vehicles follow the traffic light.

Thanks
